tests/test_edge_cases.py

```diff
@@ -1910,27 +1909,25 @@ def __init__(self, t1: T1, t2: T2):
             self.t2 = t2
 
         @classmethod
-        def __get_validators__(cls):
-            yield cls.validate
-
-        @classmethod
-        def validate(cls, v, field):
-            if not isinstance(v, cls):
-                raise TypeError('Invalid value')
-            if not field.sub_fields:
+        def __get_pydantic_core_schema__(cls, source, **kwargs):
+            schema = core_schema.is_instance_schema(cls)
+
+            args = get_args(source)
+            if not args:
+                return schema
+
+            t1_f = Validator(args[0])
+            t2_f = Validator(args[1])
+
+            def validate(v, info):
+                if not args:
+                    return v
+                # TODO: Collect these errors, rather than stopping early, and modify the loc to make the test pass
+                t1_f(v.t1)
+                t2_f(v.t2)
```

Comment:
Since we want to allow validators to raise a `ValidationError`, I wonder if the solution here is to let validators raise a `ValueError[List[ValLineError]]` (or similar, whatever we want the Python API to be)? Then they would also be able to change the `loc`. We can include the `loc` that was used to get to them via `info`.

Reply:
yeah I'm not sure what's best but I think a lot of things could work and that proposal sounds reasonable